README.md
Render 1-D barcodes to canvas, svg, or path.  Supports Code 128, 3 of 9, 2 of 5, EAN, and more.

## Why?
This barcode library has no dependencies 👍.

This barcode library specializes in its use of the canvas API.  You provide a canvas (be it in a browser or with [node-canvas](https://github.com/Automattic/node-canvas)), and then specify where the barcode should be drawn, how large it should be, and at what angle.

Other barcode libraries that I found created pngs or svgs directly and were not very flexible in the output size (i.e. no `maxWidth` option).  Rendering directly to a canvas allows for a flexible API if that's your target anyway.  You can also easily generate SVG or Path text output.

## Demos/Playground
First of all, check out the [demos](https://froatsnook.github.io/bardcode/).

## Barcode Generator

Did you ever want to paste a list of strings and download a zip file of corresponding images?  Well I did once and made this [barcode-generator](https://froatsnook.github.io/barcode-generator).  See demos for the source.

## How?
```js
// In the browser
var canvas = document.getElementById("barcode-canvas");
var g = canvas.getContext("2d");
drawBarcode(g, "Test barcode", options);
```

```js
// On the server
var bardcode = require("bardcode");
var Canvas = require("canvas");
var canvas = new Canvas(595, 842, "pdf");
var g = canvas.getContext("2d");
bardcode.drawBarcode(g, "Test barcode", options);
```

```js
// Meteor server
var Canvas = Meteor.npmRequire("canvas");
var canvas = new Canvas(595, 842, "pdf");
var g = canvas.getContext("2d");
drawBarcode(g, "Test barcode", options);
```

## Symbology
The following barcode formats are supported:

* Codabar
* Code 128 (and GS1 128)
* Code 39
* EAN-8
* EAN-13
* FIM
* ITF (interleaved 2 of 5)
* UPC-A

## API
```js
// browser+meteor
drawBarcode(g, barcodeText, options);

// node
var bardcode = require("bardcode");
bardcode.drawBarcode(g, barcodeText, options);
```

* `g` is a Canvas' 2-D graphics context or the output format.  The only supported non-canvas output format is `"svg"`.
* `barcodeText` is a string or an array (in case you want to include command characters like "FNC 1").  Allowed characters depend on the chosen symbology.  You generally don't need to include start, end, or check characters.
* `options` is an optional object with the following properties:
    * `options.type` Barcode type.  Defaults to Code 128.  Other valid options are "GS1 128", "Codabar", "Code 39", "EAN-8", "EAN-13", "FIM", "ITF" (interleaved 2 of 5), and "UPC-A".
    * `options.hasChecksum` If true, the barcode already has a checksum (which will be validated); if false, calculate and add a checksum. Defaults to false. **Currently works only for EAN-type barcodes (EAN-8, EAN-13, UPC-A).** 
    * `options.x` Where to draw barcode.  Defaults to 0.
    * `options.y` Where to draw the barcode.  Defaults to 0.
    * `options.horizontalAlign` How to align the barcode.  Defaults to "left".  Other options are "center" and "right".
    * `options.verticalAlign` How to align the barcode.  Defaults to "top".  Other options are "middle" and "bottom".
    * `options.height` Barcode height.  Defaults to 90.72.
    * `options.moduleWidth` Width of thinnest bar.  Defaults to 2.892.
    * `options.quietZoneSize` Number of moduleWidths in quiet zone on either side.  Defaults to 10.
    * `options.angle` Rotate barcode this many degrees clockwise.  Defaults to 0.
    * `options.maxWidth` Maximum barcode width (including quiet zones).  If specified, then the moduleWidth will be adjusted if necessary to make the entire barcode fit in the given width.
    * `options.width` If given, then ignore moduleWidth and maxWidth and set the moduleWidth so that the barcode will have the given width (including quiet zones).

Note: `width` and `maxWidth` refer to the unrotated barcode width.

## SVG
If the first parameter to `bardcode.drawBarcode`, `g`, is `"svg"`, then an SVG string is returned.

For the moment, the only supported options when using SVG are:
* `options.type`
* `options.height`
* `options.moduleWidth`
* `options.quietZoneSize`
* `options.maxWidth`
* `options.width`

## Path
If the first parameter to `bardcode.drawBarcode`, `g`, is `"path"`, then a path string is returned, like:

```path
M 28.920,0 l 5.784,0 l 0,90.72 l -5.784,0 Z M 37.596,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 46.272,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 60.732,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 69.408,0 l 11.568,0 l 0,90.72 l -11.568,0 Z M 83.868,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 92.544,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 98.328,0 l 5.784,0 l 0,90.72 l -5.784,0 Z M 109.896,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 124.356,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 130.140,0 l 11.568,0 l 0,90.72 l -11.568,0 Z M 147.492,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 156.168,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 164.844,0 l 11.568,0 l 0,90.72 l -11.568,0 Z M 179.304,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 187.980,0 l 11.568,0 l 0,90.72 l -11.568,0 Z M 205.332,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 211.116,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 219.792,0 l 5.784,0 l 0,90.72 l -5.784,0 Z M 234.252,0 l 8.676,0 l 0,90.72 l -8.676,0 Z M 245.820,0 l 2.892,0 l 0,90.72 l -2.892,0 Z M 251.604,0 l 5.784,0 l 0,90.72 l -5.784,0 Z
```

For the moment, the only supported options when using path are:
* `options.type`
* `options.height`
* `options.moduleWidth`
* `options.quietZoneSize`
* `options.maxWidth`
* `options.width`

## Errors
Errors are thrown on invalid input, for example when including letters in EAN barcodes.

## Installation
`node-canvas` installation causes issues for a lot of users.  So it's an optional dependency (i.e. you should Bring Your Own Canvas).  It's also not a node dependency because it's not the only output format.

For the browser:
* Include `dist/bardcode.js` or `dist/bardcode.min.js`

For node:
* `npm install canvas` (requires cairo)
* `npm install bardcode`

For meteor:
* (optional) `meteor add meteorhacks:npm` and add `"canvas": "1.1.6"` to `packages.json`
* `meteor add froatsnook:bardcode`

## Testing
Tests require meteor and zbar.

* Download zbar: http://zbar.sourceforge.net/
* `./configure --disable-video --without-qt --without-gtk --without-python`
* `make`
* Copy `zbarimg/zbarimg` and `zbar/.libs/libzbar.so*` to a new directory called `zbar` in `bardcode`'s project directory
* `meteor test-packages ./` in `bardcode`'s project directory
* Open `localhost:3000` in a browser

## Contributors
Thank you:

* [wolfgang42](https://github.com/wolfgang42)
* Tom S.
* [I Putu Susila](https://github.com/ipsusila)

## License
MIT



CHANGELOG.md
# 2.2.0 2020-08-31
* Add $/+% to Code 39 lookup table. Thanks I P Susila!

# 2.1.0 2019-05-31
* Add GS1 128 type. Thanks Tom S!

## 2.0.1 - 2018-08-14
* Fix EAN checksum check

## 2.0.0 - 2017-11-14
* Add Path output
* Make ES6 and rollup builds

## 1.2.0 - 2015-09-01
* Add `options.hasChecksum`.  Thank you wolfgang42!

## 1.1.3 - 2015-08-29
* Fix EAN checksum bug.  Thank you wolfgang42!

## 1.1.2 - 2015-03-23
* Added demos

## 1.1.1 - 2015-03-21
* Automated tests via zbar
* Fixed Code 39 rendering

## 1.1.0 - 2015-03-10
* Experimental SVG support

## 1.0.0 - 2015-02-01
* Canvas only
* Code 128
* Codabar
* Code 39
* EAN-8
* EAN-13
* FIM
* ITF (interleaved 2 of 5)
* UPC-A



demos/demos.js
(function() {

"use strict";

window.drawGrid = function(g, w, h, spacing) {
  g.strokeStyle = "black";

  for (var x = 0; x < w; x += spacing) {
    g.beginPath();
    g.moveTo(x, 0);
    g.lineTo(x, h);
    g.stroke();
  }

  for (var y = 0; y < h; y += spacing) {
    g.beginPath();
    g.moveTo(0, y);
    g.lineTo(w, y);
    g.stroke();
  }
};

var Demo = React.createClass({
  getInitialState: function() {
    return { code: this.props.children.toString().trim() };
  },

  runCode: function() {
    var code = this.state.code;
    //console.log(code);
    eval(code);
  },

  updateCode: function(e) {
    var newCode = e.target.value;
    this.setState({
      code: newCode
    });
  },

  render: function() {
    var self = this;

    // Extract the canvasId from the code.
    var canvasId = /document.getElementById\("(.*)"\)/.exec(this.state.code)[1];

    var w = /w = (\d+)/.exec(this.state.code)[1];
    var h = /h = (\d+)/.exec(this.state.code)[1];

    var codeStyle = {
        width: w + "px",
        height: h + "px"
    };
    var canvasStyle = {
        width: w + "px",
        height: h + "px"
    };

    setTimeout(function() {
      self.runCode();
    }, 0);

    return (
      <article className="demo-container">
        <div className="demo">
          <h3>{this.props.name}</h3>

          <textarea className="demo-code"
                    value={this.state.code}
                    onChange={this.updateCode}
                    autoComplete="off"
                    autoCorrect="off"
                    autoCapitalize="off"
                    spellCheck="false"
                    style={codeStyle}></textarea>

          <canvas id={canvasId}
                  className="demo-canvas"
                  width={w}
                  height={h}
                  style={canvasStyle}></canvas>
        </div>
      </article>
    );
  }
});

React.render(
  <div id="demos">

    <section>
      <h2>Introduction</h2>

      <p>By default, bardcode renders in the top left corner of the canvas.</p>

      <Demo name="demo 1 - just draw something">
{`
var w = 420, h = 200;
var canvas = document.getElementById("demo-1");
var g = canvas.getContext("2d");
g.fillStyle = "white";
g.fillRect(0, 0, w, h);
bardcode.drawBarcode(g, "test", { });
`}
      </Demo>

      <p>So why is it not at the very top left?  bardcode, by default, includes a quiet zone to the left and right of the barcode.  Since it's left aligned, there is some space on the left.  This can be configured:</p>

      <Demo name="demo 2 - no quiet zone">
{`
var w = 420, h = 200;
var canvas = document.getElementById("demo-2");
var g = canvas.getContext("2d");
g.fillStyle = "white";
g.fillRect(0, 0, w, h);
bardcode.drawBarcode(g, "test", {
    quietZoneSize: 0
});
`}
      </Demo>

      <p>You can align and position the barcode within the canvas however you want:</p>

      <Demo name="demo 3 - position and alignment">
{`
var w = 420, h = 230;
var canvas = document.getElementById("demo-3");
var g = canvas.getContext("2d");
g.fillStyle = "white";
g.fillRect(0, 0, w, h);
bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: h/2,
    horizontalAlign: "center",
    verticalAlign: "middle"
});
`}
      </Demo>

      <p>You might also find it useful to rotate the barcode.  Use <span className="mono">options.angle</span> (degrees clockwise).</p>

      <Demo name="demo 4 - rotation">
{`
var w = 420, h = 250;
var canvas = document.getElementById("demo-4");
var g = canvas.getContext("2d");
g.fillStyle = "white";
g.fillRect(0, 0, w, h);
bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: h/2,
    horizontalAlign: "center",
    verticalAlign: "middle",
    angle: 90
});
`}
      </Demo>

    </section>

    <section>
      <h2>Width and height</h2>

      <p>There are several options for controlling the width and height of rendered barcodes.</p>

      <p>Setting the <span className="mono">moduleWidth</span> sets the width of the thinnest bar.  The default value is 2.892.</p>

      <p>That is, of course, unless you specify the <span className="mono">maxWidth</span> which will decrease the <span className="mono">moduleWidth</span> (if necessary) so that the barcode will fit within it (including the quiet zone unless <span className="mono">quietZoneSize</span> is set to 0).</p>

      <p>However, both are ignored if the <span className="mono">width</span> is specfied.  The <span className="mono">moduleWidth</span> will be set to whatever value is necessary to make the barcode (plus quiet zone unless <span className="mono">quietZoneSize</span> is set to 0) have the given width.</p>

      <Demo name="demo 5 - width and height">
{`
var w = 420, h = 290;
var canvas = document.getElementById("demo-5");
var g = canvas.getContext("2d");
g.fillStyle = "white";
g.fillRect(0, 0, w, h);

drawGrid(g, w, h, 20);

bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: 0,
    moduleWidth: 2,
    quietZoneSize: 0,
    horizontalAlign: "center",
    verticalAlign: "top",
    height: 50
});
bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: 60,
    maxWidth: 2*w/3,
    horizontalAlign: "center",
    verticalAlign: "top",
    height: 50
});
bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: 120,
    maxWidth: 2*w/3,
    quietZoneSize: 0,
    horizontalAlign: "center",
    verticalAlign: "top",
    height: 50
});
bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: 180,
    width: w,
    horizontalAlign: "center",
    verticalAlign: "top",
    height: 50
});
bardcode.drawBarcode(g, "test", {
    x: w/2,
    y: 240,
    width: w,
    quietZoneSize: 0,
    horizontalAlign: "center",
    verticalAlign: "top",
    height: 50
});
`}
      </Demo>

    </section>

    <section>
      <h2>Symbologies</h2>

      <p>So far all of the barcodes have been Code-128 encoded "test".  Use the second parameter to `bardcode.drawBarcode` to change the barcode text, and set <span className="mono">options.type</span> to use a different symbology.</p>

      <p>The supported symbologies are:</p>
      <ul>
        <li>Codabar</li>
        <li>Code 128</li>
        <li>Code 39</li>
        <li>EAN-8</li>
        <li>EAN-13</li>
        <li>FIM</li>
        <li>ITF (interleaved 2 of 5)</li>
        <li>UPC-A</li>
      </ul>

      <Demo name="demo 6 - symbologies">
{`
var w = 420, h = 370;
var canvas = document.getElementById("demo-6");
var g = canvas.getContext("2d");
g.fillStyle = "white";
g.fillRect(0, 0, w, h);

var barcodes = [
  { type: "Codabar", val: "31117013206375" },
  { type: "Code 128", val: "BarCode 1" },
  { type: "Code 39", val: "0123456789" },
  { type: "EAN-8", val: "9638507" },
  { type: "EAN-13", val: "590123412345" },
  { type: "FIM", val: "C" },
  { type: "ITF", val: "04004" },
  { type: "UPC-A", val: "90123412345" }
];

var targetHeight = (h + 5)/barcodes.length;

for (var i = 0; i < barcodes.length; i++) {
  var barcode = barcodes[i];
  bardcode.drawBarcode(g, barcode.val, {
    type: barcode.type,
    x: w/2,
    y: i*targetHeight,
    maxWidth: w,
    horizontalAlign: "center",
    verticalAlign: "top",
    height: targetHeight - 5
  });
}
`}
      </Demo>

    </section>
  </div>,
  document.getElementById("content")
);

})();

